[TACHYON-606] Create MetaDataView, TierView, DirView

ooq opened the pull request:
Unit tests to be added.

apc999 commented on servers/src/main/java/tachyon/worker/block/BlockMetadataView.java:
```diff
@@ -0,0 +1,133 @@
+  /**
+   * Create a BlockMetadataView
+   *
+   * @param manager
```
javadoc please :)

apc999 commented on servers/src/main/java/tachyon/worker/block/BlockMetadataView.java:
```diff
@@ -0,0 +1,133 @@
+
+  private final BlockMetadataManager mMetadataManager;
+  private List<StorageTierView> mTierViews = new ArrayList<StorageTierView>();
+  private final List<Long> mPinnedBlocks = new ArrayList<Long>();
```
I think `List` is not optimal here, since you use `contains` which is O(n). Why not `Set`?

apc999 commented on servers/src/main/java/tachyon/worker/block/BlockMetadataView.java:
```diff
@@ -0,0 +1,133 @@
+  private final BlockMetadataManager mMetadataManager;
+  private List<StorageTierView> mTierViews = new ArrayList<StorageTierView>();
+  private final List<Long> mPinnedBlocks = new ArrayList<Long>();
+  private final List<Long> mReadingBlocks = new ArrayList<Long>();
```
Same here, why not `Set`

apc999 replied:
I prefer to rename `mReadingBlocks` to `mLockedBlocks` to be a little bit more general :)

apc999 commented on servers/src/main/java/tachyon/worker/block/BlockMetadataView.java:
```diff
@@ -0,0 +1,133 @@
+
+    // iteratively create all StorageTierViews and StorageDirViews
+    for (StorageTier tier : manager.getTiers()) {
+      StorageTierView tierView = new StorageTierView(tier, this);
```
NIT: name `tierView` is only used once, do we bother to have it?

apc999 commented on servers/src/main/java/tachyon/worker/block/BlockMetadataView.java:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  /**
+   * get pinned list
```
NIT: Capitalize the first word "get", to follow our convention.
Also the next line is unnecessary.

apc999 commented on servers/src/main/java/tachyon/worker/block/BlockMetadataView.java:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  /**
+   * get list of blocks currently being read
```
Same comment as above.

apc999 commented on servers/src/main/java/tachyon/worker/block/BlockMetadataView.java:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  /**
+   * return null if block is pinned or currently being read,
```
"return" -> "Return"

apc999 commented on servers/src/main/java/tachyon/worker/block/meta/StorageDirView.java:
```diff
@@ -0,0 +1,79 @@
+
+  public StorageDirView(StorageDir dir, StorageTierView tier, BlockMetadataView view) {
+    mDir = Preconditions.checkNotNull(dir);
+    mTier = tier;
```
why not `checkNotNull` here?

apc999 commented on servers/src/main/java/tachyon/worker/block/BlockMetadataView.java:
```diff
@@ -0,0 +1,133 @@
+/**
+ * This class exposes a narrower view of {@link BlockMetadataManager} to Evictors and Allocators.
+ */
+public class BlockMetadataView {
```
There are no setters for `mPinnedBlocks` and `mReadingBlocks` except the constructor. Do we plan to create a `BlockMetadataView` each time for each `freeSpace` with latest version of pinned and reading blocks? If yes, leave a TODO here to improve this inefficiency; if not, please add setters.

apc999 replied:
Class name is kinda confusing. It is supposed to correspond to `BlockMetadataManager`, however it looks like the counterpart of `BlockMeta` which is totally different from `BlockMetadataManager`

ooq replied:
Add a TODO.
Rename BlockMetadataView -> BlockMetadataManagerView.

apc999 commented on servers/src/main/java/tachyon/worker/block/meta/StorageDirView.java:
```diff
@@ -0,0 +1,79 @@
+
+    for (BlockMeta blockMeta : mDir.getBlocks()) {
+      long blockId = blockMeta.getBlockId();
+      if (!pinnedBlocks.contains(blockId) && !readingBlocks.contains(blockId)) {
```
The only reason you need `pinnedBlocks` and `readingBlocks` is because you need to test membership of a block Id in these two lists. Why not simply have two APIs `isBlockPinned` and `isBlockReading` in `mView`?

apc999 commented on servers/src/main/java/tachyon/worker/block/meta/StorageDirView.java:
```diff
@@ -0,0 +1,79 @@
+   * return a filtered list of block metadata,
+   * for blocks that are neither pinned or being read.
+   */
+  public List<BlockMeta> getBlocks() {
```
I prefer to have two APIs for `StorageDirView`: 
- `getBlocks` that returns all blocks (simply wrapper of `mDir.getBlocks` including pinned and reading blocks, as this may be useful to Allocator.
- `getEvictableBlocks` which returns what you are doing here which should be used by Evictor.

@calvinjia , what is your opinion?

ooq replied:
Rename `getBlocks` -> `getEvictableBlocks`.
I didn't add a `getBlocks` for all blocks yet, as none of the allocators currently need that.

apc999 commented on servers/src/main/java/tachyon/worker/block/meta/StorageDirView.java:
```diff
@@ -0,0 +1,79 @@
+  }
+
+  /**
+   * return a filtered list of block metadata,
```
"return" -> "Return"

apc999 commented on servers/src/main/java/tachyon/worker/block/meta/StorageDirView.java:
```diff
@@ -0,0 +1,79 @@
+    return mDir.getAvailableBytes();
+  }
+
+  public long getCommittedBytes() {
```
I think we shall leave this as is, and have another API `getEvictableBytes` that counts pinned and locked blocks.

ooq replied:
Not addressed yet.
Essentially we want `GreedyEvictor#canEvictBlocksFromDir` to be `return dir.getAvailableBytes() + dir.getEvictableBytes()` ?

apc999 commented on servers/src/main/java/tachyon/worker/block/meta/StorageDirView.java:
```diff
@@ -0,0 +1,79 @@
+    return mDir;
+  }
+
+  public StorageTierView getParentTier() {
```
shouldn't this be `getParentTierView`?

apc999 commented on servers/src/main/java/tachyon/worker/block/meta/StorageTierView.java:
```diff
@@ -0,0 +1,60 @@
+import tachyon.worker.block.BlockMetadataView;
+
+/**
+ * This class is a wrapper of {@link StorageTier} to provided more limited access
```
'provided' -> 'provide'

apc999 commented on servers/src/main/java/tachyon/worker/block/meta/StorageDirView.java:
```diff
@@ -0,0 +1,79 @@
+import tachyon.worker.block.BlockMetadataView;
+
+/**
+ * This class is a wrapper of {@link StorageDir} to provided more limited access
```
'provided' -> 'provide'

apc999 commented on servers/src/main/java/tachyon/worker/block/meta/StorageTierView.java:
```diff
@@ -0,0 +1,60 @@
+    return mDirViews;
+  }
+
+  public StorageDirView getDirView(int dirIndex) throws IOException {
```
Please javadoc this method especially stating when the Exception is thrown.

apc999 commented on servers/src/main/java/tachyon/worker/block/meta/StorageDirView.java:
```diff
@@ -0,0 +1,79 @@
+public class StorageDirView {
+
+  private final StorageDir mDir;
+  private final StorageTierView mTier;
```
better to rename this variable to `mTierView`

apc999 commented:
Hi @ooq , thanks for the patch. Some comments are left.

ooq commented:
@apc999 Thanks for the comments. They are great. Will pay more attentions on the details in following PRs :)

I have updated the PR with some self-improvements and addressing most of your comments. For the un-addressed ones, please find my reply above. Thanks!

apc999 commented:
Compile problem: 
[ERROR] /home/jenkins/workspace/Tachyon-Pull-Request-Builder/servers/src/main/java/tachyon/worker/block/BlockMetadataView.java:[54,34] constructor StorageTierView in class tachyon.worker.block.meta.StorageTierView cannot be applied to given types;
  required: tachyon.worker.block.meta.StorageTier,tachyon.worker.block.BlockMetadataManagerView
  found: tachyon.worker.block.meta.StorageTier,tachyon.worker.block.BlockMetadataView
  reason: actual argument tachyon.worker.block.BlockMetadataView cannot be converted to tachyon.worker.block.BlockMetadataManagerView by method invocation conversion

ooq commented:
Closing this PR, instead solving this issue in #1063
